feat: Add initial Core Web Vitals UI behind FF

rafaeelaudibert opened the pull request:
First PR on the Core Web Vitals project!

## Changes

This is a simple PR creating the page and route for our upcoming Core Web Vitals UI
There're some slight changes to some components so that our `<Filters />` component has the proper space to render on Mobile

This is hidden behind a FF for now because there isn't anything to see, but it should go behind a Beta FF soon!

![image](https://github.com/user-attachments/assets/7016e3e5-e9e4-4268-9c7a-3db40463fbae)
![image](https://github.com/user-attachments/assets/c8a0ee62-b556-4a66-9839-7c4de5ebd1c4)

https://www.loom.com/share/2ac44609fe894f9182ea76a7e0d7a20b?sid=cab54855-c9b6-40f8-bba8-579c676f89d3

## Does this work well for both Cloud and self-hosted?

Yes sir

## How did you test this code?

Manual testing, looked around the app comparing with prod to guarantee changes to `Filters` had no regressions around the app

rafaeelaudibert committed:
feat: Add new `isWindowLessThan` function to `useWindowSize`

This will allow us to more easily check whether we're on a small screen or not. Without this, we'd need to keep implementing our own breakpoint checking by using `TAILWIND_BREAKPOINTS`

rafaeelaudibert committed:
feat: Add initial Core Web Vitals UI behind FF

This is a simple commit creating the page and route for our upcoming Core Web Vitals UI

There's some slight changes to some components so that our `<Filters />` component has the proper space to render

rafaeelaudibert committed:
layout: Small fixes after changes to `PropertyFilters`

We changed how our filters work in the previous commit because we wanted a nicer layout for Web Analytics, but it added a couple of regressions in some other parts of the app, let's fix them :)

rafaeelaudibert committed:
bug: Properly change product tab when clicking on the sidebar

rafaeelaudibert committed:
bug: Add missing scene for Core Web Vitals

rafaeelaudibert committed:
bug: Fix missing export

Even though these aren't used in the app, we need them to be exported if we want the typegen to work

rafaeelaudibert committed:
feat: Add Core Web Vitals/Web Analytics breadcrumbs

rafaeelaudibert commented on frontend/src/lib/hooks/useWindowSize.js:
Moved this to a `.ts` file but git wasn't smart enough to notice that, see below

rafaeelaudibert commented on frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:
```diff
@@ -688,8 +708,31 @@ export const webAnalyticsLogic = kea<webAnalyticsLogicType>([
 
-                const pathCleaningSettingsUrl = urls.settings('project-product-analytics', 'path-cleaning')
-                const customChannelTypesUrl = urls.settings('environment-web-analytics', 'channel-type')
+                // TODO: Use actual web vitals tab, this is just a placeholder
```
Not happening on this PR, this will be a thing on the next PR

rafaeelaudibert commented on frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:
```diff
@@ -1803,78 +1853,15 @@ export const webAnalyticsLogic = kea<webAnalyticsLogicType>([
-                actions.setCompareFilter(compare_filter)
-            }
-        },
+        '/web': (pathParams, queryParams) => setFilters(actions, values, pathParams, queryParams),
+        '/web/:productTab': (pathParams, queryParams) => setFilters(actions, values, pathParams, queryParams),
```
I couldn't get this to work with a single string. I tried all sorts of regexes, but I don't think `kea` is that powerful. @mariusandra if you're keen to help here :)

mariusandra replied:
I'm not sure what you're trying to achieve, and Kea is as powerful as the user 😅. `urlToAction` takes a function that returns an object, so feel free to add whatever rows into the object that you need to.

```ts
    urlToAction(({ actions, values }) => {
        const obj = {}
        const handler = (path, search, hash) => {...}
        for (...) { obj[`/web/${product.name}`] => handler }
        return obj
    })
```

rafaeelaudibert replied:
Oh, of course, it's a function, I can just create the keys programatically, thanks

rafaeelaudibert commented on frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:
```diff
@@ -1952,3 +1939,76 @@ const checkCustomEventConversionGoalHasSessionIdsHelper = async (
+    ) {
+        actions.setDatesAndInterval(date_from, date_to, interval)
+    }
+    if (device_tab && device_tab !== values._deviceTab) {
+        actions.setDeviceTab(device_tab)
+    }
```
I would love so much if our style allowed this to be inlined, it's just so clean. One can only wonder ⭐ 

```suggestion
    if (device_tab && device_tab !== values._deviceTab) actions.setDeviceTab(device_tab)
```

mariusandra replied:
And there are others who'll consider the straightforwardness of the non-inline version as clean and elegant.

We don't support inline ifs to avoid bugs, which happen much easier this ways. While you or I might not make any, considering how this company is growing and how many people work on the codebase, we need some structure to maintain sanity :).

rafaeelaudibert replied:
> We don't support inline ifs to avoid bugs, which happen much easier this ways. 
Yeah, I can get behind, hence why I won't propose changing this, I can withstand spending 2 seconds adding curly braces if we can avoid the odd bug.

rafaeelaudibert commented on frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:
```diff
@@ -489,8 +477,36 @@ export const webAnalyticsLogic = kea<webAnalyticsLogicType>([
+        ],
     }),
     selectors(({ actions, values }) => ({
+        breadcrumbs: [
```
I'm still not sure how `breadcrumbs` work. Who's accessing these? This is like magic to me

mariusandra replied:
Whatever is your active scene, and what is that scene's logic and component, is recorded in `sceneLogic` (each scene exports its data with `export const scene: SceneExport = ...`). If the scene's logic contains a `breadcrumbs` selector, we use it to show breadcrumbs via `breadcrumbsLogic`

rafaeelaudibert replied:
> If the scene's logic contains a breadcrumbs selector

Ah ok, that's the confusing bit for me. So we're **dynamically** accessing the logic exported on `SceneExport` and then conditionally checking if they have a `breadcrumbs` selector? Sounds like typescript hell, but that makes sense. And then, if there's no selector, we simply use the `SceneExport` title as the crumb, right?

rafaeelaudibert committed:
feat: Add docs path to web analytics scene

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

mariusandra commented on frontend/src/lib/hooks/useWindowSize.ts:
You could possibly also use this: https://keajs.org/docs/plugins/window-values/

rafaeelaudibert replied:
Yes, but then I need to import `TAILWIND_BREAKPOINTS` into every logic that wants to know the screen size and reimplement the exact same logic. Using this hook should be much simpler.

mariusandra replied:
Reimplement the exact same logic? Not unless you make one common `tailwindBreakpointLogic` and use that :D

robbie-c reviewed (approved):
Sure, LGTM.

Long term we might want to separate these into difference scenes/logics, I don't know how much code we'll be sharing between the two views but even if it's a lot, that might be easier than having one scene for both.

Keen to see the product itself :)

rafaeelaudibert commented:
> Long term we might want to separate these into difference scenes/logics, I don't know how much code we'll be sharing between the two views but even if it's a lot, that might be easier than having one scene for both.

@robbie-c Yeah, fully agreed. I reused it because I thought it was easier right now - while I don't understand `kea` fully. I thought we'd share a lot of code, but we'll probably only share the header, but that's easy to solve, I can just extract that to a separate component.

robbie-c commented:
I don't think you need to worry about this too much right now though, let's focus on getting something in front of our users, even if it means creating some tech debt which we can clean up later

rafaeelaudibert committed:
refactor: Hide `setFilters` function inside `urlToAction`

This more closely explains what's this function purpose, and it also allows us to heavily simplify our types
